src/backtrace.c

```diff
@@ -0,0 +1,682 @@
+ * used to save the previous return address to the stack, and `addiu sp, sp, offset`
+ * which creates the stack frame for the current function. When we find both, we know
+ * how to get back to the previous frame. 
+ * 
+ * Notice that this also works through exceptions, as the exception handler does create
+ * a stack frame exactly like a standard function (see inthandler.S). 
+ * 
+ * Only a few functions do use a frame pointer: those that allocate a runtime-calculated
+ * amount of stack (eg: using alloca). Because of this, we actually look for usages
+ * of the frame pointer register fp, and track those as well to be able to correctly
+ * walk the stack in those cases.
+ * 
+ * Symbolization
+ * =============
+ * To symbolize the backtrace, we use a symbol table file (SYMT) that is generated
+ * by the n64sym tool during the build process. The symbol table is put into the
+ * rompak (see rompak_internal.h) and is structured in a way that can be queried
+ * directly from ROM, without even allocating memory. This is especially useful
+ * to provide backtrace in catastrophic situations where the heap is not available.
+ * 
+ * The symbol table file contains the source code references (function name, file name,
+ * line number) for a number of addresses in the ROM. Since it would be impractical to
+ * save information for all the addresses in the text segment, only special addresses
+ * are saved: in particular, those where a function call is made (ie: the address of
+ * JAL / JALR instructions), which are the ones that are commonly found in backtraces
+ * and thus need to be symbolized. In addition to these, the symbol table contains
+ * also information associated to the addresses that mark the start of each function,
+ * so that it's always possible to infer the function a certain address belongs to.
+ * 
+ * Given that not all addresses are saved, it is important to provide accurate
+ * source code references for stack frames that are interrupted by interrupts or 
+ * exceptions; in those cases, the symbolization will simply return the function name
+ * the addresses belongs to, without any source code reference.
+ * 
+ * To see more details on how the symbol table is structured in the ROM, see
+ * #symtable_header_t and the source code of the n64sym tool.
+ * 
+ */
+#include <stdint.h>
+#include <stdalign.h>
+#include <stdlib.h>
+#include <string.h>
+#include "backtrace.h"
+#include "backtrace_internal.h"
+#include "debug.h"
+#include "n64sys.h"
+#include "dma.h"
+#include "utils.h"
+#include "exception.h"
+#include "interrupt.h"
+#include "rompak_internal.h"
+
+/** @brief Enable to debug why a backtrace is wrong */
+#define BACKTRACE_DEBUG 0
+
+/** @brief Function alignment enfored by the compiler (-falign-functions). 
+ * 
+ * @note This must be kept in sync with n64.mk.
+ */
+#define FUNCTION_ALIGNMENT      32
+
+#define MAX_FILE_LEN 120        ///< Maximum length of a file name in a backtrace entry
```

Comment:
Right, this is not needed anymore, we can use VLAs to workaround this at least for the backtrace_symbols_cb case, so that it can handle any size.

For `backtrace_symbols()` (which is the standard API we implement for completeness but it's not the best one to use), since we want to avoid too many allocations, it's better to allocate once with a large enough size, so I moved the constants there.